kaolin/ops/spc/spc.py

```diff
@@ -0,0 +1,631 @@
+    to outputs :math:`Y` by the following
+
+    .. math::
+
+        Y_i = \sum_k w_k \cdot X_{n(i,k)} + b \quad\text{for}\; i \in 0,\ldots,|Y|-1.
+
+    where :math:`w_k` are weights associated with the kernel.
+
+    Args:
+        in_channels (int): the number of input channels in the
+            input tensor.
+
+        out_channels (int): the number of output channels in the
+            output tensor.
+
+        kernel_vectors (short): a python list of 3D offsets that 
+            define the shape of the kernel.
+
+        jump (int, optional): the difference between the input and
+            output levels for the convolution
+
+        bias (bool, optional): if True, the convolution layer
+            has a bias.
+
+    """
+    def __init__(self, in_channels, out_channels, kernel_vectors, jump=0, bias=True):
+        super(conv, self).__init__()
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+        self.Ksize = kernel_vectors.shape[0]
+
+        self.register_buffer('kernel_vectors', kernel_vectors)
+        self.jump = jump
+
+        self.use_mm = False  # use matrix multiplication when kernel is 1
+
+        Tensor = torch.FloatTensor
+        if self.Ksize == 1:
+            self.kernel_shape = (self.in_channels, self.out_channels)
+            self.use_mm = True
+        else:
+            self.kernel_shape = (self.Ksize, self.in_channels,
+                                 self.out_channels)
+
+        self.params = nn.Parameter(Tensor(*self.kernel_shape))
+        self.bias = nn.Parameter(Tensor(1, out_channels)) if bias else None
+        self.has_bias = bias
+
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        n = (self.in_channels) * self.Ksize
+        stdv = 1. / math.sqrt(n)
+        self.params.data.uniform_(-stdv, stdv)
+        if self.bias is not None:
+            self.bias.data.uniform_(-stdv, stdv)
+
+    def forward(self, octrees, points, pyramid, exsum, input):
+        r"""
+
+        Args:
+            octrees (`torch.ByteTensor`): structured point cloud batched octrees
+
+            points (`torch.ByteTensor`): structured point cloud batched points
+
+            pyramid (`torch.IntTensor`): structured point cloud batched pyarmids
+
+            exsum (`torch.IntTensor`): structured point cloud batched octree 
+                bit count exclusive sum
+
+            input (`torch.FloatTensor`): 
+
+        Returns:
+            (torch.Tensor): 
+                Batched (packed) output of convolution.
+
+        Example:
+            >>> # Kernel vector tensor K defined as above
+            >>> m = spc.conv(3, 16, K, jump=1)
+            ...
+            >>> # octrees, points, pyramid, exsum as above
+            >>> level = pyramid.size(2)-2
+            >>> size = pyramid[:,0,level].sum().item()
+            >>> input = torch.randn((size, 3, device=torch.device('cuda:0'))
+            >>> output = m(input)
```

Comment:
It would be great to display values of `input`, and the kernel inside the convolution layer.